### PR TITLE
feat(new): Added Azure.AKS.AuditAdmin

### LINKED ISSUE #2450: Use kube-audit-admin instead of kube-audit
### Existing rule

None

### Suggested rule

Create a new rule `Azure.AKS.AuditAdmin` to flag when the `kube-audit` log is collected with diagnostic settings.

When `kube-audit` is enabled, this can significantly increase cost for monitoring AKS clusters.

Instead enable collection for `kube-audit-admin`, which excludes the get and list audit events, but includes changes.

### Pillar

Cost Optimization

### Additional context

Related to #2249

- https://learn.microsoft.com/azure/aks/monitor-aks#resource-logs

### PR DESCRIPTION
**Pull request by BenjaminEngeset**
## PR Summary

Fixes #2450

Added Azure.AKS.AuditAdmin.

## PR Checklist

- [x] PR has a meaningful title
- [x] Summarized changes
- [x] Change is not breaking
- [x] This PR is ready to merge and is not **Work in Progress**
- **Rule changes**
  - [x] Unit tests created/ updated
  - [x] Rule documentation created/ updated
  - [x] Link to a filed issue
  - [x] [Change log](https://github.com/Azure/PSRule.Rules.Azure/blob/main/docs/CHANGELOG-v1.md) has been updated with change under unreleased section
- **Other code changes**
  - [ ] Unit tests created/ updated
  - [ ] Link to a filed issue
  - [ ] [Change log](https://github.com/Azure/PSRule.Rules.Azure/blob/main/docs/CHANGELOG-v1.md) has been updated with change under unreleased section

### CODE REVIEW

**Comment by BenjaminEngeset** [comment]
Ready for review, @BernieWhite, but I have some questions.

Could you explain what the `Microsoft.ContainerService/managedClusters/providers/diagnosticSettings` is? We have some rules implemented with this, but I don’t see examples in the MS docs using this over `Microsoft.Insights/diagnosticSettings`.

 The documentation states:

`To create a diagnostic setting for an Azure resource, add a resource of type <resource namespace>/providers/diagnosticSettings to the template.`

https://learn.microsoft.com/en-us/azure/azure-monitor/essentials/resource-manager-diagnostic-settings

However, the actual examples still use `Microsoft.Insights/diagnosticSettings`.

I would be truly grateful for any information on this.

**Comment by BernieWhite** [comment]
> Ready for review, @BernieWhite, but I have some questions.
> 
> Could you explain what the `Microsoft.ContainerService/managedClusters/providers/diagnosticSettings` is? We have some rules implemented with this, but I don’t see examples in the MS docs using this over `Microsoft.Insights/diagnosticSettings`.
> 
> The documentation states:
> 
> `To create a diagnostic setting for an Azure resource, add a resource of type <resource namespace>/providers/diagnosticSettings to the template.`
> 
> https://learn.microsoft.com/en-us/azure/azure-monitor/essentials/resource-manager-diagnostic-settings
> 
> However, the actual examples still use `Microsoft.Insights/diagnosticSettings`.
> 
> I would be truly grateful for any information on this.

Hi @BenjaminEngeset, when ARM parses type and ID for resources it gets a little interesting when dealing with extension resources.

Taking a normal child resource.

- Type: `Microsoft.ContainerService/managedClusters/agentPools` 
- ID of `/subscriptions/00000000-0000-0000-0000-000000000000/resourceGroups/rg-1/providers/Microsoft.ContainerService/managedClusters/cluster-1/agentPools/cluster-1`

After the `providers/Microsoft.ContainerService` the next segments are `type`, `name`, `type`, `name`.

Parsing this becomes:

- Provider `Microsoft.ContainerService`.
- Type `managedClusters`
- Name `cluster-1`
- Type `agentPools`
- Name `pool-1`.

Making the final type: `Microsoft.ContainerService/managedClusters/agentPools` and the final name of the child `cluster-1/cluster-1`.

---

Here's where it gets interesting.

For an extension resource of type `Microsoft.Insights/diagnosticSettings` that's nested under a `Microsoft.ContainerService/managedClusters`, the resource ID might be:

`/subscriptions/00000000-0000-0000-0000-000000000000/resourceGroups/rg-1/providers/Microsoft.ContainerService/managedClusters/cluster-1/providers/Microsoft.Insights/diagnosticSettings/logs-1`

However when ARM functionally parses this for some operations it still follows the pattern `type`, `name`, `type`, `name`, which calculates to:

- Provider `Microsoft.ContainerService`.
- Type `managedClusters`
- Name `cluster-1`
- Type `providers`
- Name `Microsoft.Insights`
- Type `diagnosticSettings`
- Name `log-1`

Making the nested type `Microsoft.ContainerService/managedClusters/providers/diagnosticSettings`.

File: docs/en/rules/Azure.AKS.AuditAdmin.md [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+## SYNOPSIS
+
+Use kube-audit-admin instead of kube-audit to capture administrative actions in AKS clusters.
+
+## DESCRIPTION
+
+Collecting resource logs for AKS clusters, particularly kube-audit logs, can be costly. Here are some recommendations to reduce data collection costs:
+
+- Disable kube-audit logging when unnecessary: Regular kube-audit logs include all API requests, which can generate a large volume of data and increase costs.
+- Enable kube-audit-admin logging: This focuses on administrative actions and excludes less critical get and list events, reducing the amount of data collected without sacrificing important security information.
```
Comment (BernieWhite) [inline docs/en/rules/Azure.AKS.AuditAdmin.md:20]:
```suggestion
Key components in a Kubernetes cluster regularly scan or check for updated Kubernetes resources against the API server.
These _get_ and _list_ operations typically occur more frequently as a Kubernetes cluster grows.

Auditing within AKS writes log events for each operation that occur against the API server.
As a result, collecting audit logs for _get_ and _list_ operations of a production AKS cluster can increase cost exponentially.

AKS provides two log categories for collecting audit logs, `kube-audit` and `kube-audit-admin`.

- `kube-audit` - Audit log data for every audit event including _get_, _list_, _create_, _update_, _delete_, _patch_, and _post_.
- `kube-audit-admin` - Is a subset of the `kube-audit` log category that excludes _get_ and list_ audit events.

In other words, both `kube-audit` and `kube-audit-admin` contain the same data except `kube-audit-admin` does not contain _get_ and _list_ events.
Changes to the cluster configuration are captured with _create_, _update_, _delete_, _patch_, and _post_ events.

By using `kube-audit-admin`, changes to resources in AKS are audited, however events relating to reading resources and configuration are not.
This significantly reduces the number of logs and overall cost for collecting and storing AKS audit events.
```

**Review by BernieWhite — CHANGES_REQUESTED** [review]
Thanks @BenjaminEngeset. Looks good. I have a few suggested changes below.

Also I think the `docs/en/rules/json` empty file can be removed.

File: src/PSRule.Rules.Azure/rules/Azure.AKS.Rule.ps1 [hunk 2]
```diff
@@ -305,6 +305,21 @@ Rule 'Azure.AKS.EphemeralOSDisk' -Ref 'AZR-000287' -Level Warning -Type 'Microso
...
+    if ($kubeAuditLogs.Count -eq 0) {
+        return $Assert.Pass()
+    }
+
+    foreach ($kubeAuditLog in $kubeAuditLogs) {
+        $Assert.HasDefaultValue($kubeAuditLog, "properties.logs[?@category == 'kube-audit'].category", 'kube-audit-admin').
+        Reason($LocalizedData.AKSAuditAdmin, $kubeAuditLog.name)
+    }
+}
+
```
[the remark this answers is not in the document]
Comment (BenjaminEngeset) [reply src/PSRule.Rules.Azure/rules/Azure.AKS.Rule.ps1:322]: You are correct, @BernieWhite. This approach is better. I over-engineered the solution, and now I see it can be resolved with less complexity. Thanks.

**Comment by BenjaminEngeset** [comment]
> > Ready for review, @BernieWhite, but I have some questions.
> > Could you explain what the `Microsoft.ContainerService/managedClusters/providers/diagnosticSettings` is? We have some rules implemented with this, but I don’t see examples in the MS docs using this over `Microsoft.Insights/diagnosticSettings`.
> > The documentation states:
> > `To create a diagnostic setting for an Azure resource, add a resource of type <resource namespace>/providers/diagnosticSettings to the template.`
> > https://learn.microsoft.com/en-us/azure/azure-monitor/essentials/resource-manager-diagnostic-settings
> > However, the actual examples still use `Microsoft.Insights/diagnosticSettings`.
> > I would be truly grateful for any information on this.
> 
> Hi @BenjaminEngeset, when ARM parses type and ID for resources it gets a little interesting when dealing with extension resources.
> 
> Taking a normal child resource.
> 
> * Type: `Microsoft.ContainerService/managedClusters/agentPools`
> * ID of `/subscriptions/00000000-0000-0000-0000-000000000000/resourceGroups/rg-1/providers/Microsoft.ContainerService/managedClusters/cluster-1/agentPools/cluster-1`
> 
> After the `providers/Microsoft.ContainerService` the next segments are `type`, `name`, `type`, `name`.
> 
> Parsing this becomes:
> 
> * Provider `Microsoft.ContainerService`.
> * Type `managedClusters`
> * Name `cluster-1`
> * Type `agentPools`
> * Name `pool-1`.
> 
> Making the final type: `Microsoft.ContainerService/managedClusters/agentPools` and the final name of the child `cluster-1/cluster-1`.
> 
> Here's where it gets interesting.
> 
> For an extension resource of type `Microsoft.Insights/diagnosticSettings` that's nested under a `Microsoft.ContainerService/managedClusters`, the resource ID might be:
> 
> `/subscriptions/00000000-0000-0000-0000-000000000000/resourceGroups/rg-1/providers/Microsoft.ContainerService/managedClusters/cluster-1/providers/Microsoft.Insights/diagnosticSettings/logs-1`
> 
> However when ARM functionally parses this for some operations it still follows the pattern `type`, `name`, `type`, `name`, which calculates to:
> 
> * Provider `Microsoft.ContainerService`.
> * Type `managedClusters`
> * Name `cluster-1`
> * Type `providers`
> * Name `Microsoft.Insights`
> * Type `diagnosticSettings`
> * Name `log-1`
> 
> Making the nested type `Microsoft.ContainerService/managedClusters/providers/diagnosticSettings`.
> 
> Functionally there is probably very few cases where you would encounter this with ARM, since Bicep does some of this under the covers but you don't need to be aware of it. It was more common with ARM templates.

Thanks @BernieWhite. This was indeed very interesting information that I highly appreciate.

If that's the case, I assume that all this happens when the deployment is sent off to the Azure Resource Manager API, because after building my Bicep to ARM, the resource type is still `Microsoft.Insights/diagnosticSettings`, so what does Bicep do under the cover here?

When exporting with `Get-AzDiagnosticSetting -ResourceId <>` the type is still `Microsoft.Insights/diagnosticSettings`. Is the *full* resource type hidden?

**Comment by BenjaminEngeset** [comment]
> Thanks @BenjaminEngeset. Looks good. I have a few suggested changes below.
> 
> Also I think the `docs/en/rules/json` empty file can be removed.

Thanks @BernieWhite. The requested changes were great. Adjusted accordingly now.

**Comment by BernieWhite** [comment]
> sume that all this happens when the deployment is sent off to the Azure Resource Manager API

This specific case occurs when nesting resources in the `resources` property of a parent resource. It doesn't happen for resources that are `scope`'d.

Also note, it is an optional syntax in ARM templates (both full and relative types are support). To my knowledge, Bicep doesn't use the relative type form when it transpiles to JSON.

Also nested resources are not commonly used, by far the most common syntax is scoped/ parented resources. 

An example is here:

https://github.com/Azure/azure-quickstart-templates/blob/4c5b993158a30a08a9084bcd9af6ea332a6d504e/demos/private-aks-cluster-with-public-dns-zone/azuredeploy.json#L1062-L1095

**Review by BernieWhite — APPROVED** [review]
Thanks @BenjaminEngeset. All good to merge.

**Comment by BenjaminEngeset** [comment]
> > If that's the case, I assume that all this happens when the deployment is sent off to the Azure Resource Manager API,
> > because after building my Bicep to ARM, the resource type is still Microsoft.Insights/diagnosticSettings, so what does Bicep do under the cover here?
> > When exporting with Get-AzDiagnosticSetting -ResourceId <> the type is still Microsoft.Insights/diagnosticSettings. Is the full resource type hidden?
> 
> @BenjaminEngeset This specific case occurs when nesting resources in the `resources` property of a parent resource. It doesn't happen for resources that are `scope`'d.
> 
> Also note, it is an optional syntax in ARM templates (both full and relative types are supported). To my knowledge, Bicep doesn't use the relative type form when it transpiles to JSON, so it is unlikely you will encounter this behaviour currently.
> 
> Also nested resources are not commonly used in Bicep, by far the most common syntax is scoped/ parented resources.
> 
> An example is here:
> 
> https://github.com/Azure/azure-quickstart-templates/blob/4c5b993158a30a08a9084bcd9af6ea332a6d504e/demos/private-aks-cluster-with-public-dns-zone/azuredeploy.json#L1062-L1095

Thanks @BernieWhite!
